framework/src/controller/jsonrpc/types.ts

```diff
@@ -13,30 +13,37 @@
  */
 
 export type ID = string | number | null;
-export type Result = string | number | boolean | object;
-export interface SuccessObject {
-	jsonrpc: string;
-	id: ID;
-	result: Result;
-}
-export interface NotificationObject {
-	jsonrpc: string;
-	method: string;
-	result?: Result;
-}
+export type JsonRpcResult = string | number | boolean | object;
+
 export interface JsonRpcError {
 	code: number;
 	message: string;
-	data?: Result;
-}
-export interface ErrorObject {
-	jsonrpc: string;
-	id: ID;
-	error: JsonRpcError;
+	data?: JsonRpcResult;
 }
+
 export interface RequestObject {
-	readonly id: string | number | null;
+	readonly id: ID;
 	readonly jsonrpc: string;
 	readonly method: string;
-	readonly params?: object;
+	readonly params?: object & { source?: string };
```

Comment:
why does json rpc request has source? maybe if you want to reuse we can extend the type and omit or add?

Reply:
When we build action from request or build request from action, no information should be lost. That's why had to add this extra params source in the params.

Reply:
the request can be json rpc or ipc/in-memory action, so what i meant was source was not required for ipc/in-memory action. we could have two separate types. since its optional you are allowing it to coexists.

Reply:
Request are always `json-rpc` no matter invoked from IPC or In-Memory. Bus only communicate in `JSON-RPC` format.